### PR TITLE
Emit events when setting unread notifications

### PR DESCRIPTION
**Pull request by gsouquet**
Paves the way for https://github.com/vector-im/element-web/issues/23192

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: src/models/room.ts [hunk 1]
```diff
@@ -127,6 +128,7 @@ export enum RoomEvent {
     OldStateUpdated = "Room.OldStateUpdated",
     CurrentStateUpdated = "Room.CurrentStateUpdated",
     HistoryImportedWithinTimeline = "Room.historyImportedWithinTimeline",
+    UnreadNotifications = "Room.UnreadNotifications",
```
Comment (turt2live) [inline src/models/room.ts:131]: it bothers me that we mix `Subject.lowerCamelCase` and `Subject.UpperCamelCase`, ftr. Personally I prefer `lowerCamelCase`, but that seems to be outvoted here.

File: src/models/room.ts [hunk 2]
```diff
@@ -186,7 +192,8 @@ export class Room extends ReadReceipt<EmittedEvents, RoomEventHandlerMap> {
     public readonly reEmitter: TypedReEmitter<EmittedEvents, RoomEventHandlerMap>;
     private txnToEvent: Record<string, MatrixEvent> = {}; // Pending in-flight requests { string: MatrixEvent }
     private notificationCounts: NotificationCount = {};
-    private threadNotifications: Map<string, NotificationCount> = new Map();
+    private readonly threadNotifications: Map<string, NotificationCount> = new Map();
```
Comment (turt2live) [inline src/models/room.ts:195]: Shorthand would be this fwiw:
```suggestion
    private readonly threadNotifications = new Map<string, NotificationCount>();
```

File: src/models/room.ts [hunk 3]
```diff
@@ -1193,8 +1206,17 @@ export class Room extends ReadReceipt<EmittedEvents, RoomEventHandlerMap> {
      * @returns The notification count, or undefined if there is no count
      *          for this type.
      */
-    public getThreadUnreadNotificationCount(threadId: string, type = NotificationCountType.Total): number | undefined {
-        return this.threadNotifications.get(threadId)?.[type];
+    public getThreadUnreadNotificationCount(threadId: string, type = NotificationCountType.Total): number {
```
Comment (turt2live) [inline src/models/room.ts:1209]: this looks like a breaking change - the function is not flagged as experimental, it's public, and it has a behavioural change.

File: src/models/room.ts [hunk 4]
```diff
@@ -1193,8 +1206,17 @@ export class Room extends ReadReceipt<EmittedEvents, RoomEventHandlerMap> {
...
+        return this.threadNotifications.get(threadId)?.[type] ?? 0;
+    }
+
+    public hasThreadUnreadNotification(): boolean {
```
Comment (turt2live) [inline src/models/room.ts:1213]: jsdoc please

**Review by turt2live — COMMENTED** [review]
generally seems fine, though the breaking change is a bit uncomfortable.

File: src/models/room.ts (hunk 3 above)
Comment (gsouquet) [reply src/models/room.ts:1209]: It should have been marked as experimental. And this is only behind a labs flag.
I believe this should not be marked as a breaking change.

File: src/models/room.ts (hunk 2 above)
Comment (gsouquet) [reply src/models/room.ts:195]: This kind of style comments should be enforced as part of the code style and not be part of reviews IMO.

**Comment by gsouquet** [comment]
@turt2live the Typescript Strict Error CI has gone a bit coocoo. Besides that I believe I've adressed all your comments

File: src/models/room.ts (hunk 3 above)
Comment (turt2live) [reply src/models/room.ts:1209]: The js-sdk doesn't have a concept of labs flags though, and we can't guarantee that people *aren't* using it. There's other breaking changes in the SDK already landed anyways, so I think we can just apply the label here and move on.

File: src/models/room.ts (hunk 2 above)
Comment (turt2live) [reply src/models/room.ts]: This was a suggestion, not a requirement.

**Review by turt2live — APPROVED** [review]

**Comment by gsouquet** [comment]
@turt2live again, this change is NOT a breaking change. This function is behind a labs flag, and on top of that it has never made it to a release as it was introduced 7 days ago.

I will be removing the label

File: src/models/room.ts [hunk 5]
```diff
@@ -1182,8 +1189,14 @@ export class Room extends ReadReceipt<EmittedEvents, RoomEventHandlerMap> {
      * @return {Number} The notification count, or undefined if there is no count
      *                  for this type.
      */
-    public getUnreadNotificationCount(type = NotificationCountType.Total): number | undefined {
-        return this.notificationCounts[type];
+    public getUnreadNotificationCount(type = NotificationCountType.Total): number {
```
Comment (turt2live) [inline src/models/room.ts:1192]: This would be a breaking change, @gsouquet

**Review by turt2live — CHANGES_REQUESTED** [review]
We should probably take this conversation OOB at this point - there are breaking changes in this PR, in at least one of the two functions, and there's no such thing as a labs flag here.

If the functions are intended to be marked experimental, please do so.

**Review by turt2live — APPROVED** [review]

**Pull request merged**